<a href="https://colab.research.google.com/github/donghwi2022/ds-sa-cp2-phase2/blob/%EC%8B%A0%EB%8F%99%ED%9C%98/data_preprocessing.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 1. 필요 라이브러리 import

In [43]:
import pandas as pd
import numpy as np
from sklearn.preprocessing import LabelEncoder

# 2. 파일 세팅 및 확인

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
pd.set_option('display.max_columns', None)

In [4]:
df = pd.read_csv("/content/drive/MyDrive/cp2_phase2_data/trans_merged.csv", index_col=0)

In [5]:
df

created_at  customer_id  \
0        2018-07-29T15:22:01.458193Z         5868   
1        2018-07-30T12:40:22.365620Z         4774   
2        2018-09-15T11:51:17.365620Z         4774   
3        2018-11-01T11:23:48.365620Z         4774   
4        2018-12-18T11:20:30.365620Z         4774   
...                              ...          ...   
1254580  2021-09-21T22:10:05.074873Z        25121   
1254581  2021-12-09T07:35:18.222046Z        34829   
1254582  2021-12-11T01:20:36.944593Z        99645   
1254583  2019-12-18T11:01:09.623382Z        87319   
1254584  2019-12-19T12:47:14.701215Z        58180   

                                   booking_id  \
0        186e2bee-0637-4710-8981-50c2d737bc42   
1        caadb57b-e808-4f94-9e96-8a7d4c9898db   
2        6000fffb-9c1a-4f4a-9296-bc8f6b622b50   
3        f5e530a7-4350-4cd1-a3bc-525b5037bcab   
4        0efc0594-dbbf-4f9a-b0b0-a488cfddf8a2   
...                                       ...   
1254580  66f02fa4-c62e-497a-8b6e-be239009b50a   
1254581  ccde6b83-ee42-4cce-b2cd-976574886eac   
1254582  5bcd3434-1dc6-44bb-ae63-fd19714cbd6c   
1254583  af7f291a-b01b-4bfd-89a9-237b16368b15   
1254584  b242e188-5f94-4448-a038-c5865842374b   

                                   session_id payment_method payment_status  \
0        3abaa6ce-e320-4e51-9469-d9f3fa328e86     Debit Card        Success   
1        2ee5ead1-f13e-4759-92df-7ff48475e970    Credit Card        Success   
2        93325fb6-eb00-4268-bb0e-6471795a0ad0            OVO        Success   
3        bcad5a61-1b67-448d-8ff4-781d67bc56e4    Credit Card        Success   
4        df1042ab-13e6-4072-b9d2-64a81974c51a    Credit Card        Success   
...                                       ...            ...            ...   
1254580  484eecde-9f4a-42de-a8df-7ce6668101f4     Debit Card         Failed   
1254581  954e1211-4d45-4d0b-ba84-53e80119592c            OVO        Success   
1254582  e6e6c108-abd0-49a2-b0a5-691117f9cfe1    Credit Card        Success   
1254583  bb6e5426-a3f3-426a-b99d-79173eaa234b    Credit Card        Success   
1254584  5e3183df-198b-40e0-b861-27ef694a832a          Gopay        Success   

         promo_amount   promo_code  shipment_fee          shipment_date_limit  \
0                1415  WEEKENDSERU         10000  2018-08-03T05:07:24.812676Z   
1                   0          NaN         10000  2018-08-03T01:29:03.415705Z   
2                   0          NaN         10000  2018-09-18T08:41:49.422380Z   
3                   0          NaN             0  2018-11-05T17:42:27.954235Z   
4                   0          NaN             0  2018-12-23T17:24:07.361785Z   
...               ...          ...           ...                          ...   
1254580             0          NaN             0  2021-09-27T00:43:40.992167Z   
1254581          5010       AZ2022         10000  2021-12-12T07:52:44.803822Z   
1254582          2181  WEEKENDSERU          5000  2021-12-12T13:28:22.848703Z   
1254583          4503       AZ2022             0  2019-12-23T22:56:46.831720Z   
1254584          4724      BUYMORE         10000  2019-12-23T06:02:05.828334Z   

         shipment_location_lat  shipment_location_long  total_amount  \
0                    -8.227893              111.969107        199832   
1                     3.013470              107.802514        155526   
2                    -2.579428              115.743885        550696   
3                    -3.602334              120.363824        271012   
4                    -3.602334              120.363824        198753   
...                        ...                     ...           ...   
1254580              -6.869535              111.077145        226820   
1254581              -8.677264              115.176947        359203   
1254582              -6.185498              106.831093       1158159   
1254583              -7.155977              110.891858        184143   
1254584              -6.155983              106.723535        365963   

         prod

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1254585 entries, 0 to 1254584
Data columns (total 40 columns):
 #   Column                  Non-Null Count    Dtype  
---  ------                  --------------    -----  
 0   created_at              1254585 non-null  object 
 1   customer_id             1254585 non-null  int64  
 2   booking_id              1254585 non-null  object 
 3   session_id              1254585 non-null  object 
 4   payment_method          1254585 non-null  object 
 5   payment_status          1254585 non-null  object 
 6   promo_amount            1254585 non-null  int64  
 7   promo_code              481139 non-null   object 
 8   shipment_fee            1254585 non-null  int64  
 9   shipment_date_limit     1254585 non-null  object 
 10  shipment_location_lat   1254585 non-null  float64
 11  shipment_location_long  1254585 non-null  float64
 12  total_amount            1254585 non-null  int64  
 13  product_id              1254585 non-null  int64  
 14  qu

In [7]:
df.isnull().sum()

created_at                     0
customer_id                    0
booking_id                     0
session_id                     0
payment_method                 0
payment_status                 0
promo_amount                   0
promo_code                773446
shipment_fee                   0
shipment_date_limit            0
shipment_location_lat          0
shipment_location_long         0
total_amount                   0
product_id                     0
quantity                       0
item_price                     0
id                             0
gender_x                       0
masterCategory                 0
subCategory                    0
articleType                    0
baseColour                   407
season                       603
year                          28
usage                       8871
productDisplayName           173
first_name                     0
last_name                      0
username                       0
email                          0
gender_y  

# 3. 컬럼별 확인 및 전처리 진행

## 1. created_at, customer_id 컬럼
- 컬럼 설명
    - created_at : 이벤트 생성 시각 
    - customer_id : 고객 ID
- 전처리 결과 <br>
: 두 컬럼 모두 개별 정보 관리를 위해 보존

## 2. booking_id, session_id 컬럼
- 컬럼 설명
    - booking_id : 예약 ID
    - session_id : 세션 ID
- 전처리 결과 <br>
: booking_id가 같은 데이터는 session_id도 같기 때문에 booking_id 컬럼은 드랍하고 session_id는 라벨링 진행


In [8]:
# 중복 데이터 확인
df[df['booking_id'].duplicated() == True]

created_at  customer_id  \
10       2019-08-10T12:14:24.365620Z         4774   
11       2019-08-10T12:14:24.365620Z         4774   
12       2019-08-10T12:14:24.365620Z         4774   
13       2019-08-10T12:14:24.365620Z         4774   
14       2019-08-10T12:14:24.365620Z         4774   
...                              ...          ...   
1254550  2019-10-14T11:54:11.214122Z        37956   
1254559  2020-05-13T19:17:26.359304Z        64772   
1254561  2020-12-02T00:10:25.359304Z        64772   
1254566  2016-11-22T13:02:52.112855Z        24508   
1254577  2016-09-22T00:46:56.728174Z        16528   

                                   booking_id  \
10       771665e9-7df2-41c8-8168-16d5cdd407a1   
11       771665e9-7df2-41c8-8168-16d5cdd407a1   
12       771665e9-7df2-41c8-8168-16d5cdd407a1   
13       771665e9-7df2-41c8-8168-16d5cdd407a1   
14       771665e9-7df2-41c8-8168-16d5cdd407a1   
...                                       ...   
1254550  59dd4a7b-b9b5-4205-974f-0640062dc827   
1254559  9003bf28-afba-4881-8edd-b05f2ce0d652   
1254561  98206a95-8095-45ec-90fe-701df7b1b654   
1254566  bb64c193-bc69-4190-bc36-cb601651e506   
1254577  e3877a13-8283-4b71-adb6-55e22ca62247   

                                   session_id payment_method payment_status  \
10       f5352cdc-0e3c-4e77-af42-645501dcded4    Credit Card        Success   
11       f5352cdc-0e3c-4e77-af42-645501dcded4    Credit Card        Success   
12       f5352cdc-0e3c-4e77-af42-645501dcded4    Credit Card        Success   
13       f5352cdc-0e3c-4e77-af42-645501dcded4    Credit Card        Success   
14       f5352cdc-0e3c-4e77-af42-645501dcded4    Credit Card        Success   
...                                       ...            ...            ...   
1254550  e25c40d6-f68c-47fd-97b8-cc6f5ab2ccfa            OVO        Success   
1254559  e46d55c9-032c-4bfc-b469-8caa3ff499d2            OVO        Success   
1254561  2b726afb-7a89-4dbc-94c5-342c77b5fd10            OVO        Success   
1254566  54406cb9-261b-4ecd-af1b-b345cbc8172c            OVO        Success   
1254577  0a166859-a3de-4b8e-8c5c-90b58e5122f1    Credit Card        Success   

         promo_amount   promo_code  shipment_fee          shipment_date_limit  \
10               9072  WEEKENDSERU         10000  2019-08-14T18:07:04.125993Z   
11               9072  WEEKENDSERU         10000  2019-08-14T18:07:04.125993Z   
12               9072  WEEKENDSERU         10000  2019-08-14T18:07:04.125993Z   
13               9072  WEEKENDSERU         10000  2019-08-14T18:07:04.125993Z   
14               9072  WEEKENDSERU         10000  2019-08-14T18:07:04.125993Z   
...               ...          ...           ...                          ...   
1254550             0          NaN         10000  2019-10-18T12:45:42.891527Z   
1254559          3981      BUYMORE             0  2020-05-19T09:40:17.878411Z   
1254561             0          NaN         10000  2020-12-07T15:49:19.440041Z   
1254566         10327      STARTUP             0  2016-11-26T13:02:57.693302Z   
1254577             0          NaN          5000  2016-09-27T13:11:08.654682Z   

         shipment_location_lat  shipment_location_long  total_amount  \
10                   -0.992004              114.438739       1295740   
11                   -0.992004              114.438739       1295740   
12                   -0.992004              114.438739       1295740   
13                   -0.992004              114.438739       1295740   
14                   -0.992004              114.438739       1295740   
...                        ...                     ...           ...   
1254550              -6.685415              107.685665        933627   
1254559              -6.939718              111.256057        365206   
1254561               0.350953               99.640838        197402   
1254566              -6.332865              106.885794       1889564   
1254577               3.347205              106.302090        531721   

         prod

In [9]:
# 중복 데이터의 고객 id값
dupli_customer_id = df[df['booking_id'].duplicated() == True].customer_id.unique()
dupli_customer_id

array([ 4774, 58191, 24915, ..., 64772, 24508, 16528])

In [10]:
# customer_id가 같으면 booking_id와 session_id가 같은지 확인
def duplicate_data_customer (customer_id) :
    return df[df['customer_id'] == customer_id]

duplicate_data_customer(64772)

created_at  customer_id  \
1254558  2020-05-13T19:17:26.359304Z        64772   
1254559  2020-05-13T19:17:26.359304Z        64772   
1254560  2020-12-02T00:10:25.359304Z        64772   
1254561  2020-12-02T00:10:25.359304Z        64772   
1254562  2021-06-22T05:50:07.359304Z        64772   
1254563  2022-01-10T13:51:41.359304Z        64772   

                                   booking_id  \
1254558  9003bf28-afba-4881-8edd-b05f2ce0d652   
1254559  9003bf28-afba-4881-8edd-b05f2ce0d652   
1254560  98206a95-8095-45ec-90fe-701df7b1b654   
1254561  98206a95-8095-45ec-90fe-701df7b1b654   
1254562  11c3a08c-8b5c-41ca-8d7c-ff01030866a6   
1254563  bee016e5-7dd9-4689-bad0-3dfdaf09ba99   

                                   session_id payment_method payment_status  \
1254558  e46d55c9-032c-4bfc-b469-8caa3ff499d2            OVO        Success   
1254559  e46d55c9-032c-4bfc-b469-8caa3ff499d2            OVO        Success   
1254560  2b726afb-7a89-4dbc-94c5-342c77b5fd10            OVO        Success   
1254561  2b726afb-7a89-4dbc-94c5-342c77b5fd10            OVO        Success   
1254562  d04c2020-fef1-4fea-868f-84a222298d59            OVO        Success   
1254563  b4a2ca36-4250-47cb-b799-89db41316b1b            OVO        Success   

         promo_amount promo_code  shipment_fee          shipment_date_limit  \
1254558          3981    BUYMORE             0  2020-05-19T09:40:17.878411Z   
1254559          3981    BUYMORE             0  2020-05-19T09:40:17.878411Z   
1254560             0        NaN         10000  2020-12-07T15:49:19.440041Z   
1254561             0        NaN         10000  2020-12-07T15:49:19.440041Z   
1254562          3598     AZ2022             0  2021-06-25T13:41:40.714248Z   
1254563             0        NaN         10000  2022-01-14T02:40:21.501322Z   

         shipment_location_lat  shipment_location_long  total_amount  \
1254558              -6.939718              111.256057        365206   
1254559              -6.939718              111.256057        365206   
1254560               0.350953               99.640838        197402   
1254561               0.350953               99.640838        197402   
1254562              -6.939718              111.256057        260386   
1254563              -4.651293              119.748660        196191   

         product_id  quantity  item_price     id gender_x masterCategory  \
1254558       46632         1      208147  46632    Women  Personal Care   
1254559       30760         1      161040  30760      Men        Apparel   
1254560       54036         1       32049  54036    Women    Accessories   
1254561       19615         1      155353  19615    Women        Apparel   
1254562       44712         1      263984  44712      Men       Footwear   
1254563       54582         1      186191  54582    Women        Apparel   

        subCategory articleType baseColour  season    year   usage  \
1254558   Fragrance   Deodorant      Green  Spring  2017.0  Casual   
1254559     Topwear      Kurtas        Red  Summer  2012.0  Ethnic   
1254560        Bags    Handbags   Charcoal  Summer  2012.0  Casual   
1254561     Topwear        Tops      Olive  Summer  2011.0  Casual   
1254562  Flip Flops  Flip Flops      Black  Summer  2010.0  Casual   
1254563     Topwear      Kurtas       Blue  Summer  2012.0  Ethnic   

                                      productDisplayName first_name  \
1254558                         Rasasi Women Romance Deo       Maya   
1254559                   Fabindia Men Printed Red Kurta       Maya   
1254560                     Kiara Women Charcoal Handbag       Maya   
1254561  United Colors of Benetton Women Solid Olive Top       Maya   
1254562               Lotto Men Black Campione Flip Flop       Maya   
1254563                Sushilas Women Printed Blue Kurta       Maya   

         last_name                              username  \
1254558  Mandasari  a96782db-f8b2-4dce-a707-169b49e1e1d1   
1254559  Mandasari  a96782db-f8b2-4dce-a707-169b49e1e1d

In [11]:
# 각 컬럼의 유니크값 개수 확인
print(df['booking_id'].nunique())
print(df['session_id'].nunique())

852584
852584


In [29]:
df_ver2 = df.drop('booking_id', axis=1)

In [30]:
encoder = LabelEncoder()
df_ver2['session_id'] = encoder.fit_transform(df_ver2['session_id'].values)
df_ver2

created_at  customer_id  session_id payment_method  \
0        2018-07-29T15:22:01.458193Z         5868      194913     Debit Card   
1        2018-07-30T12:40:22.365620Z         4774      155874    Credit Card   
2        2018-09-15T11:51:17.365620Z         4774      489413            OVO   
3        2018-11-01T11:23:48.365620Z         4774      627839    Credit Card   
4        2018-12-18T11:20:30.365620Z         4774      742846    Credit Card   
...                              ...          ...         ...            ...   
1254580  2021-09-21T22:10:05.074873Z        25121      239980     Debit Card   
1254581  2021-12-09T07:35:18.222046Z        34829      496378            OVO   
1254582  2021-12-11T01:20:36.944593Z        99645      769040    Credit Card   
1254583  2019-12-18T11:01:09.623382Z        87319      623782    Credit Card   
1254584  2019-12-19T12:47:14.701215Z        58180      312730          Gopay   

        payment_status  promo_amount   promo_code  shipment_fee  \
0              Success          1415  WEEKENDSERU         10000   
1              Success             0          NaN         10000   
2              Success             0          NaN         10000   
3              Success             0          NaN             0   
4              Success             0          NaN             0   
...                ...           ...          ...           ...   
1254580         Failed             0          NaN             0   
1254581        Success          5010       AZ2022         10000   
1254582        Success          2181  WEEKENDSERU          5000   
1254583        Success          4503       AZ2022             0   
1254584        Success          4724      BUYMORE         10000   

                 shipment_date_limit  shipment_location_lat  \
0        2018-08-03T05:07:24.812676Z              -8.227893   
1        2018-08-03T01:29:03.415705Z               3.013470   
2        2018-09-18T08:41:49.422380Z              -2.579428   
3        2018-11-05T17:42:27.954235Z              -3.602334   
4        2018-12-23T17:24:07.361785Z              -3.602334   
...                              ...                    ...   
1254580  2021-09-27T00:43:40.992167Z              -6.869535   
1254581  2021-12-12T07:52:44.803822Z              -8.677264   
1254582  2021-12-12T13:28:22.848703Z              -6.185498   
1254583  2019-12-23T22:56:46.831720Z              -7.155977   
1254584  2019-12-23T06:02:05.828334Z              -6.155983   

         shipment_location_long  total_amount  product_id  quantity  \
0                    111.969107        199832       54728         1   
1                    107.802514        155526       16193         1   
2                    115.743885        550696       53686         4   
3                    120.363824        271012       20228         1   
4                    120.363824        198753       55220         1   
...                         ...           ...         ...       ...   
1254580              111.077145        226820       30642         2   
1254581              115.176947        359203       19394         1   
1254582              106.831093       1158159       13079         4   
1254583              110.891858        184143       11312         1   
1254584              106.723535        365963       47433         1   

         item_price     id gender_x masterCategory subCategory   articleType  \
0            191247  54728      Men       Footwear       Shoes  Casual Shoes   
1            145526  16193      Men        Apparel     Topwear      Sweaters   
2            135174  53686    Women    Accessories        Bags      Handbags   
3            271012  20228    Women        Apparel     Topwear      Sweaters   
4            198753  55220    Women  Personal Care        Lips     Lip Liner   
...             ...    ...      ...            ...         ...           ...   
1254580      113410  30642      Men        Apparel     Topwear       Tshirts   
1254581   

## 3. 결제 관련 컬럼
- 컬럼 설명
    - payment_method : 결제 방식
    - payment_status : 결제 성공 여부
    - promo_amount : 할인 금액
    - promo_code : 프로모션 코드(할인쿠폰)

- 전처리 결과 <br>
: payment_status 컬럼은 결제에 실패하더라도 관심여부는 확인할 수 있기 때문에 보존, promo_code 컬럼은 할인을 하지 않으면 결측치일 수 있기 때문에 결측치가 많은 컬럼이였지만 보존

In [14]:
df_ver2[df_ver2['payment_status'] == 'Failed']

created_at  customer_id  session_id payment_method  \
149      2021-03-30T17:48:53.162570Z        58191      555029            OVO   
181      2021-12-10T19:02:44.162570Z        58191      179401            OVO   
201      2020-07-30T22:22:14.086530Z        11016      575070    Credit Card   
475      2020-12-27T03:13:12.145153Z        22811      812463            OVO   
476      2020-12-27T03:13:12.145153Z        22811      812463            OVO   
...                              ...          ...         ...            ...   
1254451  2021-01-14T00:42:32.035111Z        33739      316792    Credit Card   
1254537  2021-10-07T03:37:23.803983Z        32621       83782            OVO   
1254541  2016-07-30T07:49:56.939285Z        65187      345818            OVO   
1254564  2017-03-14T19:46:44.841192Z        97504      181020            OVO   
1254580  2021-09-21T22:10:05.074873Z        25121      239980     Debit Card   

        payment_status  promo_amount     promo_code  shipment_fee  \
149             Failed             0            NaN         10000   
181             Failed          4923         XX2022         10000   
201             Failed             0            NaN          5000   
475             Failed          5162  WEEKENDMANTAP         10000   
476             Failed          5162  WEEKENDMANTAP         10000   
...                ...           ...            ...           ...   
1254451         Failed          7508         AZ2022         10000   
1254537         Failed             0            NaN          5000   
1254541         Failed          1894    WEEKENDSERU         10000   
1254564         Failed             0            NaN         10000   
1254580         Failed             0            NaN             0   

                 shipment_date_limit  shipment_location_lat  \
149      2021-04-01T15:22:57.917698Z              -7.972558   
181      2021-12-15T23:33:45.142182Z              -8.209107   
201      2020-08-02T07:59:51.827400Z               0.078228   
475      2021-01-01T09:01:26.351985Z              -7.465412   
476      2021-01-01T09:01:26.351985Z              -7.465412   
...                              ...                    ...   
1254451  2021-01-19T13:20:40.631568Z              -6.177467   
1254537  2021-10-09T18:07:30.466350Z              -7.091944   
1254541  2016-08-02T12:25:16.032453Z              -7.965964   
1254564  2017-03-19T22:34:17.856896Z              -6.311460   
1254580  2021-09-27T00:43:40.992167Z              -6.869535   

         shipment_location_long  total_amount  product_id  quantity  \
149                  110.273691        176279       58314         1   
181                  113.605459        351303       19054         2   
201                  127.862032        221493       19166         1   
475                  112.545742        926524       25477         1   
476                  112.545742        926524       46383         1   
...                         ...           ...         ...       ...   
1254451              106.755862        292083       19983         1   
1254537              111.888872        208084       12161         1   
1254541              110.647340       3025419       58144        13   
1254564              107.144159        110861       24174         1   
1254580              111.077145        226820       30642         2   

         item_price     id gender_x masterCategory subCategory   articleType  \
149          166279  58314      Men    Accessories       Socks         Socks   
181          173113  19054    Women        Apparel     Topwear        Kurtas   
201          216493  19166      Men        Apparel     Topwear       Tshirts   
475          346072  25477      Men       Footwear       Shoes  Sports Shoes   
476          162426  46383      Men        Apparel     Topwear   Sweatshirts   
...             ...    ...      ...            ...         ...           ...   
1254451      289591  19983      Men        Apparel     Topwear    

In [15]:
# 제일 마지막 구매실패 고객이 후에 구매를 다시 시도했는지 확인
df_ver2[df_ver2['customer_id'] == 25121]

created_at  customer_id  session_id payment_method  \
1254579  2020-11-12T18:19:20.074873Z        25121      171855     Debit Card   
1254580  2021-09-21T22:10:05.074873Z        25121      239980     Debit Card   

        payment_status  promo_amount promo_code  shipment_fee  \
1254579        Success             0        NaN         10000   
1254580         Failed             0        NaN             0   

                 shipment_date_limit  shipment_location_lat  \
1254579  2020-11-18T18:06:06.493534Z              -6.286732   
1254580  2021-09-27T00:43:40.992167Z              -6.869535   

         shipment_location_long  total_amount  product_id  quantity  \
1254579              106.902229        287688       42915         1   
1254580              111.077145        226820       30642         2   

         item_price     id gender_x masterCategory subCategory articleType  \
1254579      277688  42915    Women    Accessories       Belts       Belts   
1254580      113410  30642      Men        Apparel     Topwear     Tshirts   

        baseColour  season    year   usage                 productDisplayName  \
1254579       Grey  Summer  2012.0  Casual  French Connection Women Grey Belt   
1254580      White  Summer  2012.0  Casual     Nike Men Printed White T-shirt   

        first_name  last_name                              username  \
1254579    Himawan  Kurniawan  ef7a59f0-f5d4-42b5-85f5-82e093c733cf   
1254580    Himawan  Kurniawan  ef7a59f0-f5d4-42b5-85f5-82e093c733cf   

                                                     email gender_y  \
1254579  ef7a59f0_f5d4_42b5_85f5_82e093c733cf@startupca...        M   
1254580  ef7a59f0_f5d4_42b5_85f5_82e093c733cf@startupca...        M   

          birthdate device_type                             device_id  \
1254579  1993-05-19     Android  4e168860-5c6f-4fa3-a0a7-5b10ff9b0bab   
1254580  1993-05-19     Android  4e168860-5c6f-4fa3-a0a7-5b10ff9b0bab   

        device_version  home_location_lat  home_location_long home_location  \
1254579  Android 2.2.1          -6.869535          111.077145   Jawa Tengah   
1254580  Android 2.2.1          -6.869535          111.077145   Jawa Tengah   

        home_country first_join_date  
1254579    Indonesia      2020-11-12  
1254580    Indonesia      2020-11-12

## 4. 배송 관련 컬럼
- 컬럼 설명
    - shipment_fee : 배송 비용
    - shipment_date_limit : 배송까지 최대 시각(=늦어도 이때까지 도착 예정)
    - shipment_location_lat : 배송지역 위도
    - shipment_location_long : 배송지역 경도

- 전처리 결과 <br>
: 4가지 컬럼 모두 보존

## 5. 금액 관련 컬럼
- 컬럼 설명
    - total_amount : 총 비용
    - product_id : 상품id
    - quantity : 주문 수량
    - item_price : 상품 가격

- 전처리 결과 <br>
: "상품 가격 x 상품 수량 - 할인 가격 + 배송료 = 총 비용"에 맞는 데이터와 맞지 않는 데이터 분리 예정

In [18]:
# "상품 가격 x 상품 수량 - 할인 가격 + 배송료 = 총 비용"이 맞지 않는 데이터 확인
df_ver2[df_ver2['quantity']*df_ver2['item_price']+df_ver2['shipment_fee']-df_ver2['promo_amount'] != df_ver2['total_amount']]

created_at  customer_id  session_id payment_method  \
9        2019-08-10T12:14:24.365620Z         4774      816750    Credit Card   
10       2019-08-10T12:14:24.365620Z         4774      816750    Credit Card   
11       2019-08-10T12:14:24.365620Z         4774      816750    Credit Card   
12       2019-08-10T12:14:24.365620Z         4774      816750    Credit Card   
13       2019-08-10T12:14:24.365620Z         4774      816750    Credit Card   
...                              ...          ...         ...            ...   
1254561  2020-12-02T00:10:25.359304Z        64772      144382            OVO   
1254565  2016-11-22T13:02:52.112855Z        24508      279869            OVO   
1254566  2016-11-22T13:02:52.112855Z        24508      279869            OVO   
1254576  2016-09-22T00:46:56.728174Z        16528       33802    Credit Card   
1254577  2016-09-22T00:46:56.728174Z        16528       33802    Credit Card   

        payment_status  promo_amount   promo_code  shipment_fee  \
9              Success          9072  WEEKENDSERU         10000   
10             Success          9072  WEEKENDSERU         10000   
11             Success          9072  WEEKENDSERU         10000   
12             Success          9072  WEEKENDSERU         10000   
13             Success          9072  WEEKENDSERU         10000   
...                ...           ...          ...           ...   
1254561        Success             0          NaN         10000   
1254565        Success         10327      STARTUP             0   
1254566        Success         10327      STARTUP             0   
1254576        Success             0          NaN          5000   
1254577        Success             0          NaN          5000   

                 shipment_date_limit  shipment_location_lat  \
9        2019-08-14T18:07:04.125993Z              -0.992004   
10       2019-08-14T18:07:04.125993Z              -0.992004   
11       2019-08-14T18:07:04.125993Z              -0.992004   
12       2019-08-14T18:07:04.125993Z              -0.992004   
13       2019-08-14T18:07:04.125993Z              -0.992004   
...                              ...                    ...   
1254561  2020-12-07T15:49:19.440041Z               0.350953   
1254565  2016-11-26T13:02:57.693302Z              -6.332865   
1254566  2016-11-26T13:02:57.693302Z              -6.332865   
1254576  2016-09-27T13:11:08.654682Z               3.347205   
1254577  2016-09-27T13:11:08.654682Z               3.347205   

         shipment_location_long  total_amount  product_id  quantity  \
9                    114.438739       1295740       53051         1   
10                   114.438739       1295740       46352         1   
11                   114.438739       1295740        6352         1   
12                   114.438739       1295740       26620         1   
13                   114.438739       1295740       22849         1   
...                         ...           ...         ...       ...   
1254561               99.640838        197402       19615         1   
1254565              106.885794       1889564       13410         1   
1254566              106.885794       1889564        6068         5   
1254576              106.302090        531721       20764         1   
1254577              106.302090        531721       25724         1   

         item_price     id gender_x masterCategory subCategory   articleType  \
9            292052  53051    Women        Apparel       Saree        Sarees   
10           241945  46352      Men        Apparel     Topwear        Shirts   
11           297248   6352   Unisex       Footwear      Sandal       Sandals   
12           191138  26620      Men       Footwear       Shoes  Sports Shoes   
13            78559  22849      Men       Footwear       Shoes  Casual Shoes   
...             ...    ...      ...            ...         ...           ...   
1254561      155353  19615    Women        Apparel     Topwear          Tops   
1254565   

In [17]:
# "상품 가격 x 상품 수량 - 할인 가격 + 배송료 = 총 비용"이 맞는 데이터 확인
df_ver2[df_ver2['quantity']*df_ver2['item_price']+df_ver2['shipment_fee']-df_ver2['promo_amount'] == df_ver2['total_amount']]

created_at  customer_id  session_id payment_method  \
0        2018-07-29T15:22:01.458193Z         5868      194913     Debit Card   
1        2018-07-30T12:40:22.365620Z         4774      155874    Credit Card   
2        2018-09-15T11:51:17.365620Z         4774      489413            OVO   
3        2018-11-01T11:23:48.365620Z         4774      627839    Credit Card   
4        2018-12-18T11:20:30.365620Z         4774      742846    Credit Card   
...                              ...          ...         ...            ...   
1254580  2021-09-21T22:10:05.074873Z        25121      239980     Debit Card   
1254581  2021-12-09T07:35:18.222046Z        34829      496378            OVO   
1254582  2021-12-11T01:20:36.944593Z        99645      769040    Credit Card   
1254583  2019-12-18T11:01:09.623382Z        87319      623782    Credit Card   
1254584  2019-12-19T12:47:14.701215Z        58180      312730          Gopay   

        payment_status  promo_amount   promo_code  shipment_fee  \
0              Success          1415  WEEKENDSERU         10000   
1              Success             0          NaN         10000   
2              Success             0          NaN         10000   
3              Success             0          NaN             0   
4              Success             0          NaN             0   
...                ...           ...          ...           ...   
1254580         Failed             0          NaN             0   
1254581        Success          5010       AZ2022         10000   
1254582        Success          2181  WEEKENDSERU          5000   
1254583        Success          4503       AZ2022             0   
1254584        Success          4724      BUYMORE         10000   

                 shipment_date_limit  shipment_location_lat  \
0        2018-08-03T05:07:24.812676Z              -8.227893   
1        2018-08-03T01:29:03.415705Z               3.013470   
2        2018-09-18T08:41:49.422380Z              -2.579428   
3        2018-11-05T17:42:27.954235Z              -3.602334   
4        2018-12-23T17:24:07.361785Z              -3.602334   
...                              ...                    ...   
1254580  2021-09-27T00:43:40.992167Z              -6.869535   
1254581  2021-12-12T07:52:44.803822Z              -8.677264   
1254582  2021-12-12T13:28:22.848703Z              -6.185498   
1254583  2019-12-23T22:56:46.831720Z              -7.155977   
1254584  2019-12-23T06:02:05.828334Z              -6.155983   

         shipment_location_long  total_amount  product_id  quantity  \
0                    111.969107        199832       54728         1   
1                    107.802514        155526       16193         1   
2                    115.743885        550696       53686         4   
3                    120.363824        271012       20228         1   
4                    120.363824        198753       55220         1   
...                         ...           ...         ...       ...   
1254580              111.077145        226820       30642         2   
1254581              115.176947        359203       19394         1   
1254582              106.831093       1158159       13079         4   
1254583              110.891858        184143       11312         1   
1254584              106.723535        365963       47433         1   

         item_price     id gender_x masterCategory subCategory   articleType  \
0            191247  54728      Men       Footwear       Shoes  Casual Shoes   
1            145526  16193      Men        Apparel     Topwear      Sweaters   
2            135174  53686    Women    Accessories        Bags      Handbags   
3            271012  20228    Women        Apparel     Topwear      Sweaters   
4            198753  55220    Women  Personal Care        Lips     Lip Liner   
...             ...    ...      ...            ...         ...           ...   
1254580      113410  30642      Men        Apparel     Topwear       Tshirts   
1254581   

## 6. 상품 정보 관련 컬럼
- 컬럼 설명
    - id : 상품 id
    - gender_x : 상품 타겟 성별
    - masterCategory : 최상위 분류
    - subCategory : 서브 분류
    - articleType : 상품 종류
    - baseColour : 기본 색상 (결측치 O)
    - season : 적합 계절 (결측치 O)
    - year : 상품 출시 년도 (결측치 O)
    - usage : 사용 방식 (결측치 O)
    - productDisplayName : 상품 이름 (결측치 O)

- 전처리 결과 <br>
    - 해당 컬럼이 결측값이더라도 다른 컬럼을 사용할 수 있고, 결측치가 많지 않아(제일 많은 컬럼이 8871개) 결측치 데이터를 제거하지 않고 EDA 진행 → 후에 모델링 진행시에 결측치 처리 예정
    - id 컬럼은 product_id 컬럼과 같기 때문에 drop
    - 상품 타겟 성별 컬럼은 "gender_x → product_gender"로 변경
    - 기본 색상 컬럼은 "baseColour → baseColor"로 변경

In [31]:
df_ver2.drop(columns=['id'], inplace=True)
df_ver3 = df_ver2.rename(columns={'baseColour':'baseColor', 'gender_x':'product_gender'})

In [32]:
df_ver3

created_at  customer_id  session_id payment_method  \
0        2018-07-29T15:22:01.458193Z         5868      194913     Debit Card   
1        2018-07-30T12:40:22.365620Z         4774      155874    Credit Card   
2        2018-09-15T11:51:17.365620Z         4774      489413            OVO   
3        2018-11-01T11:23:48.365620Z         4774      627839    Credit Card   
4        2018-12-18T11:20:30.365620Z         4774      742846    Credit Card   
...                              ...          ...         ...            ...   
1254580  2021-09-21T22:10:05.074873Z        25121      239980     Debit Card   
1254581  2021-12-09T07:35:18.222046Z        34829      496378            OVO   
1254582  2021-12-11T01:20:36.944593Z        99645      769040    Credit Card   
1254583  2019-12-18T11:01:09.623382Z        87319      623782    Credit Card   
1254584  2019-12-19T12:47:14.701215Z        58180      312730          Gopay   

        payment_status  promo_amount   promo_code  shipment_fee  \
0              Success          1415  WEEKENDSERU         10000   
1              Success             0          NaN         10000   
2              Success             0          NaN         10000   
3              Success             0          NaN             0   
4              Success             0          NaN             0   
...                ...           ...          ...           ...   
1254580         Failed             0          NaN             0   
1254581        Success          5010       AZ2022         10000   
1254582        Success          2181  WEEKENDSERU          5000   
1254583        Success          4503       AZ2022             0   
1254584        Success          4724      BUYMORE         10000   

                 shipment_date_limit  shipment_location_lat  \
0        2018-08-03T05:07:24.812676Z              -8.227893   
1        2018-08-03T01:29:03.415705Z               3.013470   
2        2018-09-18T08:41:49.422380Z              -2.579428   
3        2018-11-05T17:42:27.954235Z              -3.602334   
4        2018-12-23T17:24:07.361785Z              -3.602334   
...                              ...                    ...   
1254580  2021-09-27T00:43:40.992167Z              -6.869535   
1254581  2021-12-12T07:52:44.803822Z              -8.677264   
1254582  2021-12-12T13:28:22.848703Z              -6.185498   
1254583  2019-12-23T22:56:46.831720Z              -7.155977   
1254584  2019-12-23T06:02:05.828334Z              -6.155983   

         shipment_location_long  total_amount  product_id  quantity  \
0                    111.969107        199832       54728         1   
1                    107.802514        155526       16193         1   
2                    115.743885        550696       53686         4   
3                    120.363824        271012       20228         1   
4                    120.363824        198753       55220         1   
...                         ...           ...         ...       ...   
1254580              111.077145        226820       30642         2   
1254581              115.176947        359203       19394         1   
1254582              106.831093       1158159       13079         4   
1254583              110.891858        184143       11312         1   
1254584              106.723535        365963       47433         1   

         item_price product_gender masterCategory subCategory   articleType  \
0            191247            Men       Footwear       Shoes  Casual Shoes   
1            145526            Men        Apparel     Topwear      Sweaters   
2            135174          Women    Accessories        Bags      Handbags   
3            271012          Women        Apparel     Topwear      Sweaters   
4            198753          Women  Personal Care        Lips     Lip Liner   
...             ...            ...            ...         ...           ...   
1254580      113410            Men        Apparel     Topwear       Tshirts   
1254581      35421

## 7. 고객 정보 관련 컬럼
- 컬럼 설명
    - first_name : 고객 이름
    - last_name : 고객 성
    - username : 고객 이름
    - email : 고객 이메일
    - gender_y : 고객 성별
    - birthdate : 고객 생일

- 전처리 결과 <br>
    - 고객 이름 ~ 고객 이메일에 해당하는 컬럼은 앞의 customer_id로 대체 가능하므로 모두 drop
    - "gender_y → customer_gender"로 컬럼명 변경

In [33]:
df_ver3.drop(columns=['first_name', 'last_name', 'username', 'email'], inplace=True)
df_ver4 = df_ver3.rename(columns={'gender_y':'customer_gender'})

In [34]:
df_ver4

created_at  customer_id  session_id payment_method  \
0        2018-07-29T15:22:01.458193Z         5868      194913     Debit Card   
1        2018-07-30T12:40:22.365620Z         4774      155874    Credit Card   
2        2018-09-15T11:51:17.365620Z         4774      489413            OVO   
3        2018-11-01T11:23:48.365620Z         4774      627839    Credit Card   
4        2018-12-18T11:20:30.365620Z         4774      742846    Credit Card   
...                              ...          ...         ...            ...   
1254580  2021-09-21T22:10:05.074873Z        25121      239980     Debit Card   
1254581  2021-12-09T07:35:18.222046Z        34829      496378            OVO   
1254582  2021-12-11T01:20:36.944593Z        99645      769040    Credit Card   
1254583  2019-12-18T11:01:09.623382Z        87319      623782    Credit Card   
1254584  2019-12-19T12:47:14.701215Z        58180      312730          Gopay   

        payment_status  promo_amount   promo_code  shipment_fee  \
0              Success          1415  WEEKENDSERU         10000   
1              Success             0          NaN         10000   
2              Success             0          NaN         10000   
3              Success             0          NaN             0   
4              Success             0          NaN             0   
...                ...           ...          ...           ...   
1254580         Failed             0          NaN             0   
1254581        Success          5010       AZ2022         10000   
1254582        Success          2181  WEEKENDSERU          5000   
1254583        Success          4503       AZ2022             0   
1254584        Success          4724      BUYMORE         10000   

                 shipment_date_limit  shipment_location_lat  \
0        2018-08-03T05:07:24.812676Z              -8.227893   
1        2018-08-03T01:29:03.415705Z               3.013470   
2        2018-09-18T08:41:49.422380Z              -2.579428   
3        2018-11-05T17:42:27.954235Z              -3.602334   
4        2018-12-23T17:24:07.361785Z              -3.602334   
...                              ...                    ...   
1254580  2021-09-27T00:43:40.992167Z              -6.869535   
1254581  2021-12-12T07:52:44.803822Z              -8.677264   
1254582  2021-12-12T13:28:22.848703Z              -6.185498   
1254583  2019-12-23T22:56:46.831720Z              -7.155977   
1254584  2019-12-23T06:02:05.828334Z              -6.155983   

         shipment_location_long  total_amount  product_id  quantity  \
0                    111.969107        199832       54728         1   
1                    107.802514        155526       16193         1   
2                    115.743885        550696       53686         4   
3                    120.363824        271012       20228         1   
4                    120.363824        198753       55220         1   
...                         ...           ...         ...       ...   
1254580              111.077145        226820       30642         2   
1254581              115.176947        359203       19394         1   
1254582              106.831093       1158159       13079         4   
1254583              110.891858        184143       11312         1   
1254584              106.723535        365963       47433         1   

         item_price product_gender masterCategory subCategory   articleType  \
0            191247            Men       Footwear       Shoes  Casual Shoes   
1            145526            Men        Apparel     Topwear      Sweaters   
2            135174          Women    Accessories        Bags      Handbags   
3            271012          Women        Apparel     Topwear      Sweaters   
4            198753          Women  Personal Care        Lips     Lip Liner   
...             ...            ...            ...         ...           ...   
1254580      113410            Men        Apparel     Topwear       Tshirts   
1254581      35421

## 8. 기기 정보 관련 컬럼
- 컬럼 설명
    - device_type : 디바이스 타입
    - device_id : 디바이스 id
    - device_version : 디바이스 버전

- 전처리 결과 <br>
: device_id 컬럼에 라벨 인코딩 적용

In [37]:
encoder = LabelEncoder()
df_ver4['device_id'] = encoder.fit_transform(df_ver4['device_id'].values)
df_ver4

created_at  customer_id  session_id payment_method  \
0        2018-07-29T15:22:01.458193Z         5868      194913     Debit Card   
1        2018-07-30T12:40:22.365620Z         4774      155874    Credit Card   
2        2018-09-15T11:51:17.365620Z         4774      489413            OVO   
3        2018-11-01T11:23:48.365620Z         4774      627839    Credit Card   
4        2018-12-18T11:20:30.365620Z         4774      742846    Credit Card   
...                              ...          ...         ...            ...   
1254580  2021-09-21T22:10:05.074873Z        25121      239980     Debit Card   
1254581  2021-12-09T07:35:18.222046Z        34829      496378            OVO   
1254582  2021-12-11T01:20:36.944593Z        99645      769040    Credit Card   
1254583  2019-12-18T11:01:09.623382Z        87319      623782    Credit Card   
1254584  2019-12-19T12:47:14.701215Z        58180      312730          Gopay   

        payment_status  promo_amount   promo_code  shipment_fee  \
0              Success          1415  WEEKENDSERU         10000   
1              Success             0          NaN         10000   
2              Success             0          NaN         10000   
3              Success             0          NaN             0   
4              Success             0          NaN             0   
...                ...           ...          ...           ...   
1254580         Failed             0          NaN             0   
1254581        Success          5010       AZ2022         10000   
1254582        Success          2181  WEEKENDSERU          5000   
1254583        Success          4503       AZ2022             0   
1254584        Success          4724      BUYMORE         10000   

                 shipment_date_limit  shipment_location_lat  \
0        2018-08-03T05:07:24.812676Z              -8.227893   
1        2018-08-03T01:29:03.415705Z               3.013470   
2        2018-09-18T08:41:49.422380Z              -2.579428   
3        2018-11-05T17:42:27.954235Z              -3.602334   
4        2018-12-23T17:24:07.361785Z              -3.602334   
...                              ...                    ...   
1254580  2021-09-27T00:43:40.992167Z              -6.869535   
1254581  2021-12-12T07:52:44.803822Z              -8.677264   
1254582  2021-12-12T13:28:22.848703Z              -6.185498   
1254583  2019-12-23T22:56:46.831720Z              -7.155977   
1254584  2019-12-23T06:02:05.828334Z              -6.155983   

         shipment_location_long  total_amount  product_id  quantity  \
0                    111.969107        199832       54728         1   
1                    107.802514        155526       16193         1   
2                    115.743885        550696       53686         4   
3                    120.363824        271012       20228         1   
4                    120.363824        198753       55220         1   
...                         ...           ...         ...       ...   
1254580              111.077145        226820       30642         2   
1254581              115.176947        359203       19394         1   
1254582              106.831093       1158159       13079         4   
1254583              110.891858        184143       11312         1   
1254584              106.723535        365963       47433         1   

         item_price product_gender masterCategory subCategory   articleType  \
0            191247            Men       Footwear       Shoes  Casual Shoes   
1            145526            Men        Apparel     Topwear      Sweaters   
2            135174          Women    Accessories        Bags      Handbags   
3            271012          Women        Apparel     Topwear      Sweaters   
4            198753          Women  Personal Care        Lips     Lip Liner   
...             ...            ...            ...         ...           ...   
1254580      113410            Men        Apparel     Topwear       Tshirts   
1254581      35421

## 9. 고객 집 위치 정보 관련 컬럼
- 컬럼 설명
    - home_location_lat : 집 위도
    - home_location_long : 집 경도
    - home_location : 집 위치
    - home_country : 집이 속한 나라

- 전처리 결과 <br>
: home_country 컬럼의 값이 인도네시아밖에 존재하지 않으므로 drop

In [38]:
df_ver4['home_country'].unique()

array(['Indonesia'], dtype=object)

In [40]:
df_ver5 = df_ver4.drop(columns=['home_country'])
df_ver5

created_at  customer_id  session_id payment_method  \
0        2018-07-29T15:22:01.458193Z         5868      194913     Debit Card   
1        2018-07-30T12:40:22.365620Z         4774      155874    Credit Card   
2        2018-09-15T11:51:17.365620Z         4774      489413            OVO   
3        2018-11-01T11:23:48.365620Z         4774      627839    Credit Card   
4        2018-12-18T11:20:30.365620Z         4774      742846    Credit Card   
...                              ...          ...         ...            ...   
1254580  2021-09-21T22:10:05.074873Z        25121      239980     Debit Card   
1254581  2021-12-09T07:35:18.222046Z        34829      496378            OVO   
1254582  2021-12-11T01:20:36.944593Z        99645      769040    Credit Card   
1254583  2019-12-18T11:01:09.623382Z        87319      623782    Credit Card   
1254584  2019-12-19T12:47:14.701215Z        58180      312730          Gopay   

        payment_status  promo_amount   promo_code  shipment_fee  \
0              Success          1415  WEEKENDSERU         10000   
1              Success             0          NaN         10000   
2              Success             0          NaN         10000   
3              Success             0          NaN             0   
4              Success             0          NaN             0   
...                ...           ...          ...           ...   
1254580         Failed             0          NaN             0   
1254581        Success          5010       AZ2022         10000   
1254582        Success          2181  WEEKENDSERU          5000   
1254583        Success          4503       AZ2022             0   
1254584        Success          4724      BUYMORE         10000   

                 shipment_date_limit  shipment_location_lat  \
0        2018-08-03T05:07:24.812676Z              -8.227893   
1        2018-08-03T01:29:03.415705Z               3.013470   
2        2018-09-18T08:41:49.422380Z              -2.579428   
3        2018-11-05T17:42:27.954235Z              -3.602334   
4        2018-12-23T17:24:07.361785Z              -3.602334   
...                              ...                    ...   
1254580  2021-09-27T00:43:40.992167Z              -6.869535   
1254581  2021-12-12T07:52:44.803822Z              -8.677264   
1254582  2021-12-12T13:28:22.848703Z              -6.185498   
1254583  2019-12-23T22:56:46.831720Z              -7.155977   
1254584  2019-12-23T06:02:05.828334Z              -6.155983   

         shipment_location_long  total_amount  product_id  quantity  \
0                    111.969107        199832       54728         1   
1                    107.802514        155526       16193         1   
2                    115.743885        550696       53686         4   
3                    120.363824        271012       20228         1   
4                    120.363824        198753       55220         1   
...                         ...           ...         ...       ...   
1254580              111.077145        226820       30642         2   
1254581              115.176947        359203       19394         1   
1254582              106.831093       1158159       13079         4   
1254583              110.891858        184143       11312         1   
1254584              106.723535        365963       47433         1   

         item_price product_gender masterCategory subCategory   articleType  \
0            191247            Men       Footwear       Shoes  Casual Shoes   
1            145526            Men        Apparel     Topwear      Sweaters   
2            135174          Women    Accessories        Bags      Handbags   
3            271012          Women        Apparel     Topwear      Sweaters   
4            198753          Women  Personal Care        Lips     Lip Liner   
...             ...            ...            ...         ...           ...   
1254580      113410            Men        Apparel     Topwear       Tshirts   
1254581      35421

## 10. first_join_date 컬럼 및 시계열 컬럼
- 컬럼 설명
    - first_join_date : 첫 가입 날짜

- 전처리 결과 <br>
    - birthdate, first_join_date 컬럼 : 기존방식 유지
    - year 컬럼 : 정수형 변경이 불가능하여 기존방식 유지
    - created_at, shipment_date_limit : 초 단위까지만 나오도록 길이 제한

In [79]:
df_ver5[df_ver5['year'].isnull() != True]

created_at  customer_id  session_id payment_method  \
0        2018-07-29T15:22:01.458193Z         5868      194913     Debit Card   
1        2018-07-30T12:40:22.365620Z         4774      155874    Credit Card   
2        2018-09-15T11:51:17.365620Z         4774      489413            OVO   
3        2018-11-01T11:23:48.365620Z         4774      627839    Credit Card   
4        2018-12-18T11:20:30.365620Z         4774      742846    Credit Card   
...                              ...          ...         ...            ...   
1254580  2021-09-21T22:10:05.074873Z        25121      239980     Debit Card   
1254581  2021-12-09T07:35:18.222046Z        34829      496378            OVO   
1254582  2021-12-11T01:20:36.944593Z        99645      769040    Credit Card   
1254583  2019-12-18T11:01:09.623382Z        87319      623782    Credit Card   
1254584  2019-12-19T12:47:14.701215Z        58180      312730          Gopay   

        payment_status  promo_amount   promo_code  shipment_fee  \
0              Success          1415  WEEKENDSERU         10000   
1              Success             0          NaN         10000   
2              Success             0          NaN         10000   
3              Success             0          NaN             0   
4              Success             0          NaN             0   
...                ...           ...          ...           ...   
1254580         Failed             0          NaN             0   
1254581        Success          5010       AZ2022         10000   
1254582        Success          2181  WEEKENDSERU          5000   
1254583        Success          4503       AZ2022             0   
1254584        Success          4724      BUYMORE         10000   

                 shipment_date_limit  shipment_location_lat  \
0        2018-08-03T05:07:24.812676Z              -8.227893   
1        2018-08-03T01:29:03.415705Z               3.013470   
2        2018-09-18T08:41:49.422380Z              -2.579428   
3        2018-11-05T17:42:27.954235Z              -3.602334   
4        2018-12-23T17:24:07.361785Z              -3.602334   
...                              ...                    ...   
1254580  2021-09-27T00:43:40.992167Z              -6.869535   
1254581  2021-12-12T07:52:44.803822Z              -8.677264   
1254582  2021-12-12T13:28:22.848703Z              -6.185498   
1254583  2019-12-23T22:56:46.831720Z              -7.155977   
1254584  2019-12-23T06:02:05.828334Z              -6.155983   

         shipment_location_long  total_amount  product_id  quantity  \
0                    111.969107        199832       54728         1   
1                    107.802514        155526       16193         1   
2                    115.743885        550696       53686         4   
3                    120.363824        271012       20228         1   
4                    120.363824        198753       55220         1   
...                         ...           ...         ...       ...   
1254580              111.077145        226820       30642         2   
1254581              115.176947        359203       19394         1   
1254582              106.831093       1158159       13079         4   
1254583              110.891858        184143       11312         1   
1254584              106.723535        365963       47433         1   

         item_price product_gender masterCategory subCategory   articleType  \
0            191247            Men       Footwear       Shoes  Casual Shoes   
1            145526            Men        Apparel     Topwear      Sweaters   
2            135174          Women    Accessories        Bags      Handbags   
3            271012          Women        Apparel     Topwear      Sweaters   
4            198753          Women  Personal Care        Lips     Lip Liner   
...             ...            ...            ...         ...           ...   
1254580      113410            Men        Apparel     Topwear       Tshirts   
1254581      35421

In [84]:
# 컬럼 길이 제한
df_ver5['created_at'] = df_ver5['created_at'].str[:19]
df_ver5['shipment_date_limit'] = df_ver5['shipment_date_limit'].str[:19]
df_ver5

created_at  customer_id  session_id payment_method  \
0        2018-07-29T15:22:01         5868      194913     Debit Card   
1        2018-07-30T12:40:22         4774      155874    Credit Card   
2        2018-09-15T11:51:17         4774      489413            OVO   
3        2018-11-01T11:23:48         4774      627839    Credit Card   
4        2018-12-18T11:20:30         4774      742846    Credit Card   
...                      ...          ...         ...            ...   
1254580  2021-09-21T22:10:05        25121      239980     Debit Card   
1254581  2021-12-09T07:35:18        34829      496378            OVO   
1254582  2021-12-11T01:20:36        99645      769040    Credit Card   
1254583  2019-12-18T11:01:09        87319      623782    Credit Card   
1254584  2019-12-19T12:47:14        58180      312730          Gopay   

        payment_status  promo_amount   promo_code  shipment_fee  \
0              Success          1415  WEEKENDSERU         10000   
1              Success             0          NaN         10000   
2              Success             0          NaN         10000   
3              Success             0          NaN             0   
4              Success             0          NaN             0   
...                ...           ...          ...           ...   
1254580         Failed             0          NaN             0   
1254581        Success          5010       AZ2022         10000   
1254582        Success          2181  WEEKENDSERU          5000   
1254583        Success          4503       AZ2022             0   
1254584        Success          4724      BUYMORE         10000   

         shipment_date_limit  shipment_location_lat  shipment_location_long  \
0        2018-08-03T05:07:24              -8.227893              111.969107   
1        2018-08-03T01:29:03               3.013470              107.802514   
2        2018-09-18T08:41:49              -2.579428              115.743885   
3        2018-11-05T17:42:27              -3.602334              120.363824   
4        2018-12-23T17:24:07              -3.602334              120.363824   
...                      ...                    ...                     ...   
1254580  2021-09-27T00:43:40              -6.869535              111.077145   
1254581  2021-12-12T07:52:44              -8.677264              115.176947   
1254582  2021-12-12T13:28:22              -6.185498              106.831093   
1254583  2019-12-23T22:56:46              -7.155977              110.891858   
1254584  2019-12-23T06:02:05              -6.155983              106.723535   

         total_amount  product_id  quantity  item_price product_gender  \
0              199832       54728         1      191247            Men   
1              155526       16193         1      145526            Men   
2              550696       53686         4      135174          Women   
3              271012       20228         1      271012          Women   
4              198753       55220         1      198753          Women   
...               ...         ...       ...         ...            ...   
1254580        226820       30642         2      113410            Men   
1254581        359203       19394         1      354213            Men   
1254582       1158159       13079         4      288835            Men   
1254583        184143       11312         1      188646            Men   
1254584        365963       47433         1      360687          Women   

        masterCategory subCategory   articleType baseColor  season    year  \
0             Footwear       Shoes  Casual Shoes     Black  Summer  2012.0   
1              Apparel     Topwear      Sweaters     Green    Fall  2011.0   
2          Accessories        Bags      Handbags    Purple  Summer  2012.0   
3              Apparel     Topwear      Sweaters       Red    Fall  2011.0   
4        Personal Care        Lips     Lip Liner     Brown  Spring  2017.0   
...                ...         ...        

## 11. csv파일로 저장

변경 내역들
- booking_id가 같은 데이터는 session_id도 같기 때문에 booking_id 컬럼은 드랍
- session_id 컬럼 라벨인코딩 진행
- id 컬럼은 product_id 컬럼과 같기 때문에 drop
- 컬럼명 "gender_x → product_gender" 변경
- 컬럼명 "baseColour → baseColor" 변경
- "first_name, last_name, username, email" 컬럼은 앞의 customer_id로 대체 가능하므로 모두 drop
- "gender_y → customer_gender"로 컬럼명 변경
- device_id 컬럼에 라벨 인코딩 진행
- home_country 컬럼의 값이 인도네시아밖에 존재하지 않으므로 drop
- "created_at, shipment_date_limit" 컬럼 초 단위까지만 나오도록 길이 제한

컬럼 설명 <br>
- created_at : 이벤트 발생시각
- customer_id : 고객 id
- session_id : 세션 id
- payment_method : 결제 방식
- payment_status : 결제 성공 여부
- promo_amount : 할인 금액
- promo_code : 프로모션 코드(할인쿠폰)
- shipment_fee : 배송 비용
- shipment_date_limit : 배송까지 최대 시각(언제까지는 도착할 것이다 느낌)
- shipment_location_lat : 배송지역 위도
- shipment_location_long : 배송지역 경도
- total_amount : 총 비용    
- product_id : 상품 id    
- quantity : 주문 수량
- item_price : 상품 가격 
- product_gender : 상품 타겟 성별
- masterCategory : 최상위 분류
- subCategory : 서브 분류     
- articleType : 상품 종류  
- baseColor : 기본 색상   
- season : 적합 계절    
- year : 상품 출시 년도 
- usage : 사용 방식(어느 복장인지)
- productDisplayName : 상품 이름 
- customer_gender : 고객 성별 
- birthdate : 고객 생일   
- device_type : 디바이스 타입  
- device_id : 디바이스 id  
- device_version : 디바이스 버전
- home_location_lat : 집 위도   
- home_location_long : 집 경도  
- home_location : 집 위치   
- first_join_date : 첫 가입 날짜

In [89]:
df_ver5.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1254585 entries, 0 to 1254584
Data columns (total 33 columns):
 #   Column                  Non-Null Count    Dtype  
---  ------                  --------------    -----  
 0   created_at              1254585 non-null  object 
 1   customer_id             1254585 non-null  int64  
 2   session_id              1254585 non-null  int64  
 3   payment_method          1254585 non-null  object 
 4   payment_status          1254585 non-null  object 
 5   promo_amount            1254585 non-null  int64  
 6   promo_code              481139 non-null   object 
 7   shipment_fee            1254585 non-null  int64  
 8   shipment_date_limit     1254585 non-null  object 
 9   shipment_location_lat   1254585 non-null  float64
 10  shipment_location_long  1254585 non-null  float64
 11  total_amount            1254585 non-null  int64  
 12  product_id              1254585 non-null  int64  
 13  quantity                1254585 non-null  int64  
 14  it

In [88]:
df_ver5.to_csv('fashion_log.csv')

# 4. 데이터를 두 종류로 분류할 경우
- "상품 가격 x 상품 수량 - 할인 가격 + 배송료 = 총 비용"에 맞는 데이터
- "상품 가격 x 상품 수량 - 할인 가격 + 배송료 = 총 비용"에 맞지 않는 데이터

In [86]:
df_ver5[df_ver5['quantity']*df_ver5['item_price']+df_ver5['shipment_fee']-df_ver5['promo_amount'] != df_ver5['total_amount']]

created_at  customer_id  session_id payment_method  \
9        2019-08-10T12:14:24         4774      816750    Credit Card   
10       2019-08-10T12:14:24         4774      816750    Credit Card   
11       2019-08-10T12:14:24         4774      816750    Credit Card   
12       2019-08-10T12:14:24         4774      816750    Credit Card   
13       2019-08-10T12:14:24         4774      816750    Credit Card   
...                      ...          ...         ...            ...   
1254561  2020-12-02T00:10:25        64772      144382            OVO   
1254565  2016-11-22T13:02:52        24508      279869            OVO   
1254566  2016-11-22T13:02:52        24508      279869            OVO   
1254576  2016-09-22T00:46:56        16528       33802    Credit Card   
1254577  2016-09-22T00:46:56        16528       33802    Credit Card   

        payment_status  promo_amount   promo_code  shipment_fee  \
9              Success          9072  WEEKENDSERU         10000   
10             Success          9072  WEEKENDSERU         10000   
11             Success          9072  WEEKENDSERU         10000   
12             Success          9072  WEEKENDSERU         10000   
13             Success          9072  WEEKENDSERU         10000   
...                ...           ...          ...           ...   
1254561        Success             0          NaN         10000   
1254565        Success         10327      STARTUP             0   
1254566        Success         10327      STARTUP             0   
1254576        Success             0          NaN          5000   
1254577        Success             0          NaN          5000   

         shipment_date_limit  shipment_location_lat  shipment_location_long  \
9        2019-08-14T18:07:04              -0.992004              114.438739   
10       2019-08-14T18:07:04              -0.992004              114.438739   
11       2019-08-14T18:07:04              -0.992004              114.438739   
12       2019-08-14T18:07:04              -0.992004              114.438739   
13       2019-08-14T18:07:04              -0.992004              114.438739   
...                      ...                    ...                     ...   
1254561  2020-12-07T15:49:19               0.350953               99.640838   
1254565  2016-11-26T13:02:57              -6.332865              106.885794   
1254566  2016-11-26T13:02:57              -6.332865              106.885794   
1254576  2016-09-27T13:11:08               3.347205              106.302090   
1254577  2016-09-27T13:11:08               3.347205              106.302090   

         total_amount  product_id  quantity  item_price product_gender  \
9             1295740       53051         1      292052          Women   
10            1295740       46352         1      241945            Men   
11            1295740        6352         1      297248         Unisex   
12            1295740       26620         1      191138            Men   
13            1295740       22849         1       78559            Men   
...               ...         ...       ...         ...            ...   
1254561        197402       19615         1      155353          Women   
1254565       1889564       13410         1      243311            Men   
1254566       1889564        6068         5      331316            Men   
1254576        531721       20764         1      362206          Women   
1254577        531721       25724         1      164515            Men   

        masterCategory subCategory   articleType  baseColor  season    year  \
9              Apparel       Saree        Sarees     Maroon    Fall  2012.0   
10             Apparel     Topwear        Shirts   Burgundy  Summer  2012.0   
11            Footwear      Sandal       Sandals  Navy Blue  Summer  2011.0   
12            Footwear       Shoes  Sports Shoes      Black  Summer  2012.0   
13            Footwear       Shoes  Casual Shoes      Khaki  Summer  2012.0   
...                ...         ...  

In [87]:
df_ver5[df_ver5['quantity']*df_ver5['item_price']+df_ver5['shipment_fee']-df_ver5['promo_amount'] == df_ver5['total_amount']]

created_at  customer_id  session_id payment_method  \
0        2018-07-29T15:22:01         5868      194913     Debit Card   
1        2018-07-30T12:40:22         4774      155874    Credit Card   
2        2018-09-15T11:51:17         4774      489413            OVO   
3        2018-11-01T11:23:48         4774      627839    Credit Card   
4        2018-12-18T11:20:30         4774      742846    Credit Card   
...                      ...          ...         ...            ...   
1254580  2021-09-21T22:10:05        25121      239980     Debit Card   
1254581  2021-12-09T07:35:18        34829      496378            OVO   
1254582  2021-12-11T01:20:36        99645      769040    Credit Card   
1254583  2019-12-18T11:01:09        87319      623782    Credit Card   
1254584  2019-12-19T12:47:14        58180      312730          Gopay   

        payment_status  promo_amount   promo_code  shipment_fee  \
0              Success          1415  WEEKENDSERU         10000   
1              Success             0          NaN         10000   
2              Success             0          NaN         10000   
3              Success             0          NaN             0   
4              Success             0          NaN             0   
...                ...           ...          ...           ...   
1254580         Failed             0          NaN             0   
1254581        Success          5010       AZ2022         10000   
1254582        Success          2181  WEEKENDSERU          5000   
1254583        Success          4503       AZ2022             0   
1254584        Success          4724      BUYMORE         10000   

         shipment_date_limit  shipment_location_lat  shipment_location_long  \
0        2018-08-03T05:07:24              -8.227893              111.969107   
1        2018-08-03T01:29:03               3.013470              107.802514   
2        2018-09-18T08:41:49              -2.579428              115.743885   
3        2018-11-05T17:42:27              -3.602334              120.363824   
4        2018-12-23T17:24:07              -3.602334              120.363824   
...                      ...                    ...                     ...   
1254580  2021-09-27T00:43:40              -6.869535              111.077145   
1254581  2021-12-12T07:52:44              -8.677264              115.176947   
1254582  2021-12-12T13:28:22              -6.185498              106.831093   
1254583  2019-12-23T22:56:46              -7.155977              110.891858   
1254584  2019-12-23T06:02:05              -6.155983              106.723535   

         total_amount  product_id  quantity  item_price product_gender  \
0              199832       54728         1      191247            Men   
1              155526       16193         1      145526            Men   
2              550696       53686         4      135174          Women   
3              271012       20228         1      271012          Women   
4              198753       55220         1      198753          Women   
...               ...         ...       ...         ...            ...   
1254580        226820       30642         2      113410            Men   
1254581        359203       19394         1      354213            Men   
1254582       1158159       13079         4      288835            Men   
1254583        184143       11312         1      188646            Men   
1254584        365963       47433         1      360687          Women   

        masterCategory subCategory   articleType baseColor  season    year  \
0             Footwear       Shoes  Casual Shoes     Black  Summer  2012.0   
1              Apparel     Topwear      Sweaters     Green    Fall  2011.0   
2          Accessories        Bags      Handbags    Purple  Summer  2012.0   
3              Apparel     Topwear      Sweaters       Red    Fall  2011.0   
4        Personal Care        Lips     Lip Liner     Brown  Spring  2017.0   
...                ...         ...        